File: my-folder/0022-generate-parentheses/solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtrack(res, "", n, n);
        return res;
    }

    void backtrack(List<String> list, String current, int open, int close) {
        if (open == 0 && close == 0) {
            list.add(current);
            return;
        }
        if (open > 0) {
            backtrack(list, current+'(', open-1, close);
        }
        if (close > open) {
            backtrack(list, current + ')', open, close - 1);
        }
    }
}

File: my-folder/0022-generate-parentheses/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # Keep track of the number of open and close parentheses
        # If both equal to n we add it the list
        # We only add the close parenthese after the opening ones
        # If the number of open parentheses is smaller than n, add a new one
        # If the number of close parentheses is smaller than the opening ones, add a new one
        self.res = []
        def helper(current, open, close):
            if open == 0 and close == 0:
                self.res.append(current[:])
                return
            if open > 0:
                helper(current + '(', open-1, close)
            if open < close:
                helper(current + ')', open, close - 1)
        helper("", n, n)
        return self.res
